Makefile

```diff
@@ -2,58 +2,153 @@
+lab: env
+	poetry run jupyter lab --ip=0.0.0.0 --no-browser --ServerApp.token=deadbeef
+
+# Build the documentation website.
+docs: env-docs $(shell find docs -type f) mkdocs.yml
+	poetry run mkdocs build --clean
+	rm -Rf site/overrides
 
 # Serve the documentation website.
-serve: .conda/docs
-	$(CONDA) .conda/docs
-	mkdocs serve -a 127.0.0.1:8000
+docs-serve: env-docs
+	poetry run mkdocs serve -a 127.0.0.1:8000
 
 # Serve the documentation website.
-serve-debug: .conda/docs
-	$(CONDA) .conda/docs
-	mkdocs serve -a 127.0.0.1:8000 --verbose
+docs-serve-debug: env-docs
+	poetry run mkdocs serve -a 127.0.0.1:8000 --verbose
 
 # The --dirty flag makes mkdocs not regenerate everything when change is detected but also seems to
 # break references.
-serve-dirty: .conda/docs
-	$(CONDA) .conda/docs
-	mkdocs serve --dirty -a 127.0.0.1:8000
+docs-serve-dirty: env-docs
+	poetry run mkdocs serve --dirty -a 127.0.0.1:8000
 
-# Build the documentation website.
-site: .conda/docs $(shell find docs -type f) mkdocs.yml
-	$(CONDA) .conda/docs
-	mkdocs build --clean
-	rm -Rf site/overrides
 
-upload: .conda/docs $(shell find docs -type f) mkdocs.yml
-	$(CONDA) .conda/docs
-	mkdocs gh-deploy
+docs-upload: env-docs $(shell find docs -type f) mkdocs.yml
+	poetry run mkdocs gh-deploy
 
 # Check that links in the documentation are valid. Requires the lychee tool.
-linkcheck: site
+docs-linkcheck: site
 	lychee "site/**/*.html"
+
+# Start the trubot slack app.
+trubot:
+	poetry run python -u examples/trubot/trubot.py
+
+# Run a test with the optional flag set, meaning @optional_test decorated tests
+# are run.
+
+coverage:
+	ALLOW_OPTIONAL_ENV_VAR=true pytest --rootdir=. tests/* --cov src --cov-report html
+
+# Runs required tests
+test-%-required: env-required
+	make test-$*
+
+# Runs required tests, but allows optional dependencies to be installed.
+test-%-allow-optional: env
+	ALLOW_OPTIONAL_ENV_VAR=true make test-$*
+
+# Requires the full optional environment to be set up.
+test-%-optional: env-optional
+	TEST_OPTIONAL=true make test-$*
+
+# Run the unit tests, those in the tests/unit. They are run in the CI pipeline frequently.
+test-unit:
+	poetry run pytest --rootdir=. tests/unit/*
+
+# Tests in the e2e folder make use of possibly costly endpoints. They
+# are part of only the less frequently run release tests.
+test-e2e:
+	poetry run pytest --rootdir=. tests/e2e/*
+
+# Runs the notebook test
+test-notebook:
+	poetry run pytest --rootdir=. tests/docs_notebooks/*
+
+# Release Steps:
+## Step: Clean repo:
+clean-dashboard:
+	rm -rf src/dashboard/*.egg-info
+
+clean: clean-dashboard
+	git clean --dry-run -fxd
+	@read -p "Do you wish to remove these files? (y/N)" -n 1 -r
+	echo
+	if [[ $$REPLY =~ ^[Yy]$$ ]]; then \
+		git clean -fxd; \
+	fi;
+
+## Step: Build wheels
+build-dashboard: env clean-dashboard
+	poetry run python -m build src/dashboard -o $(REPO_ROOT)/dist/trulens-dashboard;
+
+build: $(POETRY_DIRS)
+	for dir in $(POETRY_DIRS); do \
```

Comment:
btw I think if you just left the `.ONESHELL` thing you wouldn't need all these `\`.